File: Java/20. Timer/app/src/main/java/com/example/timer/MainActivity.java
package com.example.timer;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.os.CountDownTimer;
import android.os.Handler;
import android.util.Log;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

//        Handler handler = new Handler();
//        Runnable runnable = new Runnable() {
//            @Override
//            public void run() {
//                Log.d("runnable", "2 sec passed");
//                handler.postDelayed(this, 2000);
//            }
//        };
//
//        handler.post(runnable);

        CountDownTimer myTimer = new CountDownTimer(10000, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {
                Log.d("runnable", "left " + millisUntilFinished/1000);
            }

            @Override
            public void onFinish() {
                Log.d("runnable", "END");
            }
        };

        myTimer.start();
    }
}

File: MVVM_Counter/app/src/main/java/com/example/mvvm_counter/MainActivity.kt
package com.example.mvvm_counter

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.example.mvvm_counter.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {
    private lateinit var viewModel: CounterViewModel
    private lateinit var binding: ActivityMainBinding


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        binding = DataBindingUtil.setContentView(
            this, R.layout.activity_main
        )

        viewModel = ViewModelProvider(this).get(CounterViewModel::class.java)

//        val currentCount = viewModel.currentCount
        binding.increaseButton.setOnClickListener {
            viewModel.increaseValue()
//            currentCount.value = currentCount.value?.plus(1)
        }

        binding.decreaseButton.setOnClickListener {
//            currentCount.value = currentCount.value?.minus(1)
            viewModel.decreaseValue()
        }

        val counterObserver = Observer<Int> { newCounter ->
            binding.counter = newCounter
        }
        viewModel.currentCount.observe(this, counterObserver)
    }

}